kafka/consume.go

```diff
@@ -79,39 +85,66 @@ func NewConsumer(log logrus.FieldLogger, conf Config, opts ...ConfigOpt) (*Confl
-	}
-
 	cc := &ConfluentConsumer{
-		c:         consumer,
-		conf:      conf,
-		log:       log,
-		partition: part,
+		c:    consumer,
+		conf: conf,
+		log:  log,
 	}
 
-	if cc.partition < 0 {
+	if cc.conf.Consumer.Partition == nil {
 		// if we are not reading from a specific partition, we subscribe
-		cc.configOnce.Do(func() {
-			cc.log.WithField("kafka_topic", cc.conf.Topic).Debug("Subscribing to Kafka topic")
-			cc.rebalanceHandlerMutex.Lock()
-			defer cc.rebalanceHandlerMutex.Unlock()
-			if err := cc.c.Subscribe(cc.conf.Topic, cc.rebalanceHandler); err != nil {
-				err = errors.Wrap(err, "error subscribing to topic")
-			}
-		})
+		cc.setupRebalanceHandler(cc.conf.Consumer.InitialOffset)
+		cc.log.WithField("kafka_topic", cc.conf.Topic).Debug("Subscribing to Kafka topic")
+		cc.rebalanceHandlerMutex.Lock()
+		defer cc.rebalanceHandlerMutex.Unlock()
+		if serr := cc.c.Subscribe(cc.conf.Topic, cc.rebalanceHandler); serr != nil {
+			err = errors.Wrap(serr, "error subscribing to topic")
+		}
 	} else {
 		// if we are reading from a specific partition, we assign
-		cc.configOnce.Do(func() {
-			parts := []kafkalib.TopicPartition{
-				kafkalib.TopicPartition{
-					Topic:     &cc.conf.Topic,
-					Partition: cc.partition,
-				},
+		tps := []kafkalib.TopicPartition{
+			kafkalib.TopicPartition{
+				Topic:     &cc.conf.Topic,
+				Partition: *cc.conf.Consumer.Partition,
+			},
+		}
+		// Set the partition if a key is set to determine the partition
+		if cc.conf.Consumer.PartitionKey != "" && cc.conf.Consumer.PartitionerAlgorithm != "" {
+			cc.SetPartitionByKey(cc.conf.Consumer.PartitionKey, cc.conf.Consumer.PartitionerAlgorithm)
+		}
+		err = cc.c.Assign(tps)
+
+		// handle rebalance events for consumer assigned to a specific partition
+		// - NOTE if the partition is set and the partition key is not set we have no way
+		//   of knowing where to assign the consumer to in the case of a rebalance
+		cc.eventChan = cc.c.Events()
+		go func(cc *ConfluentConsumer) {
+			for ev := range cc.eventChan {
+				log := cc.log.WithField("kafka_event", ev.String())
+				switch e := ev.(type) {
+				case kafkalib.RevokedPartitions:
+					// check if we are assigned to this partition
+					revokedParts := e.Partitions
+					revoked := false
+					for _, part := range revokedParts {
+						if part.Partition == *cc.conf.Consumer.Partition && *part.Topic == cc.conf.Topic {
+							revoked = true
+							break
+						}
+					}
+					if revoked {
+						cc.log.WithField("kafka_event", e.String()).Debug("Unassigning Kafka partitions after rebalance")
+						if err := cc.c.Unassign(); err != nil {
+							log.WithError(err).Error("failed unassigning current Kafka partitions after rebalance")
+						}
+						// if we know the partition key we can reassign
+						if cc.conf.Consumer.PartitionKey != "" && cc.conf.Consumer.PartitionerAlgorithm != "" {
+							cc.SetPartitionByKey(cc.conf.Consumer.PartitionKey, cc.conf.Consumer.PartitionerAlgorithm)
+						}
```

Comment:
What is going to happen if I am assigned to partition 1 and then that partition goes away?

Reply:
The underlying c library will return a Kafka error.  There is no way to really know a partition to switch to.

Reply:
will it cause the consumer to shutdown?